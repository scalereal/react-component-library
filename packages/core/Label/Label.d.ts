import * as React from 'react';
import { WithStyle } from '../../utils';
import { LabelProps } from './types';
export declare const labelSizes: LabelProps['labelSize'][];
export declare const labelWeights: LabelProps['labelWeight'][];
declare const Label: React.FC<LabelProps> & WithStyle;
export default Label;
