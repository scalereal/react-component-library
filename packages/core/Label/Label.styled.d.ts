import { LabelProps } from './types';
export declare const Label: import("styled-components").StyledComponent<"label", any, LabelProps, never>;
