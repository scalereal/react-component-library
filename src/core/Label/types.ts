import React, { RefObject  } from 'react';
import { AnyStyledComponent } from 'styled-components';
import { LabelSizes } from "../../theme/core/label/types"
export interface WithStyle {
    Style?: AnyStyledComponent;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type HTMLPropsReact<T> = Omit<React.HTMLProps<T>, 'ref' | 'as'>;

export interface HTMLProps<T> extends HTMLPropsReact<T> {
    ref?: ((instance: T) => void) | RefObject<T>;
}
export interface WithThemeProp {
    // eslint-disable-next-line no-undef
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface LabelProps extends HTMLProps<HTMLLabelElement>, WithThemeProp   {
    /** Shows pointer while hovering on it*/
    showPointer?: boolean;
    /** Shows red color asterisk*/
    required?: boolean;
    /** Label Color */
    labelColor?: string;
    /** Label Size */
    labelSize: LabelSizes;
    /** Label Hover Color*/
    hoverColor?:string;
}
