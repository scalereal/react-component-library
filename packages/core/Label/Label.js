var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import * as Styled from './Label.styled';
var Label = React.memo(React.forwardRef(function (props, ref) { return (React.createElement(Styled.Label, __assign({ ref: ref }, props), props.children)); }));
Label.defaultProps = {
    required: false
};
Label.displayName = 'Label';
Label.Style = Styled.Label;
export default Label;
