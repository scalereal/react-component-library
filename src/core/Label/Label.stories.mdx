import Label, { labelSizes, labelWeights } from './Label';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Label"
    component={Label}
    argTypes={{
        labelColor: {
            control: 'color'
        },
        labelSize: {
            control: {
                type: 'select',
                options: Object.values(labelSizes)
            }
        },
        labelWeight: {
            control: {
                type: 'select',
                options: Object.values(labelWeights)
            }
        }
    }}
/>

# Label

Displays a Label

export const Template = ({ labelColor, required, showPointer, labelSize, labelWeight, lineHeight, upperCase, ...args }) => {
    return (
        <Label
            labelColor={labelColor}
            required={required}
            showPointer={showPointer}
            labelSize={labelSize}
            labelWeight={labelWeight}
            lineHeight={lineHeight}
            uppercase={upperCase}
            {...args}
        >
            Lorem Ipsum is simply dummy text
        </Label>
    );
};

<Preview>
    <Story
        name="Label"
        args={{
            labelSize: 'S'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Label labelColor="blue">Lorem Ipsum is simply dummy text</Label>
        <Label labelColor="red">Lorem Ipsum is simply dummy text</Label>
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Label labelColor="blue" labelSize="XS">
            Lorem Ipsum is simply dummy text
        </Label>
        <Label labelColor="blue" labelSize="S">
            Lorem Ipsum is simply dummy text
        </Label>
        <Label labelColor="blue" labelSize="M">
            Lorem Ipsum is simply dummy text
        </Label>
        <Label labelColor="blue" labelSize="L">
            Lorem Ipsum is simply dummy text
        </Label>
        <Label labelColor="blue" labelSize="XL">
            Lorem Ipsum is simply dummy text
        </Label>
    </div>
</Preview>

### Props

<Props story="Label" />
