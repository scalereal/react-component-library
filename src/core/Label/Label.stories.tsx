import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Label from "./Label"


export default {
    title: 'Components/Core/Label',
    component: Label,
    argTypes: {
        labelColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        }
    }
} as Meta;

const Template:Story = ({ labelColor, required, showPointer, hoverColor, labelSize, ...args}) => {
    return <Label labelColor={labelColor} required={required} showPointer={showPointer} hoverColor={hoverColor} labelSize={labelSize} {...args}>
    Lorem Ipsum is simply dummy text
    </Label>
}

export const Default = Template.bind({});
Default.args = {
    labelSize:'S'
}


