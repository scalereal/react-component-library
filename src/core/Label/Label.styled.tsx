import styled,{ css } from 'styled-components'
import { LabelProps } from './types'
import { defaultTheme } from '../../theme'

const asterisk = () => css`
    ::after {
        color: red;
        content: ' *';
    }
`;
const uppercase = () => css`
    text-transform: uppercase;
`;
const getTextStyle = ({ ...props }: LabelProps) => {
    const { labelColor, lineHeight, labelSize, labelWeight } = props;
    return css`
        margin: 0;
        color: ${({ theme }) => (labelColor ? labelColor : theme.label.defaults.labelColor)};
        line-height: ${({ theme }) => (lineHeight ? lineHeight : theme.label.defaults.lineHeight)};
        font-size: ${({ theme }) => (theme.label.sizes[labelSize].fontSize)};
        font-weight: ${({ theme })=> (labelWeight ? theme.label.weights[labelWeight].weight : theme.label.defaults.labelWeight)};
        font-family: ${({ theme }) => (theme.label.fontFamily)};
    `;
};

export const Label = styled('label')<LabelProps>`
    ${getTextStyle}
    ${({ required }) => required && asterisk()}
    ${props => props.uppercase && uppercase()};
    cursor: ${({ showPointer }) => (showPointer === undefined ? 'default' : showPointer ? 'pointer' : 'not-allowed')};
`;
Label.displayName = "Label"
Label.defaultProps = {
    theme:defaultTheme,
}
