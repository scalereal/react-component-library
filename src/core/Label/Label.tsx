import React, { FC } from 'react';
import * as Styled from "./Label.styled"
import { LabelProps, WithStyle } from './types';

const Label: FC<LabelProps> & WithStyle = React.memo(
    React.forwardRef((props, ref) => (
        <Styled.Label ref={ref} {...props}>
            {props.children}
        </Styled.Label>
    ))
);

Label.defaultProps = {
    required: false
};
Label.displayName = 'Label';
Label.Style = Styled.Label;

export default Label
