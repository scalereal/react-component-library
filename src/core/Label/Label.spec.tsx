import { render } from '@testing-library/react';
import * as React from 'react';
import Label from './Label';

describe('Label', () => {
    it('should render correctly with default props', () => {
        const { container } = render(<Label labelSize="S">Label</Label>);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with all the given props', () => {
        const { container } = render(
            <Label required showPointer uppercase labelColor="pink" labelSize="S" lineHeight="1" labelWeight="medium">
                Demo Label
            </Label>
        );
        expect(container).toMatchSnapshot();
    });
});
