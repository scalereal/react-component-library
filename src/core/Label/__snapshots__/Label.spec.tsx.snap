// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label should render correctly with all the given props 1`] = `
<div>
  <label

    class="sc-bdfBwQ bbGbjQ"
    required=""
  >
    Demo Label
  </label>
</div>
`;

exports[`Label should render correctly with default props 1`] = `
<div>
  <label
    class="sc-bdfBwQ cRGQsz"
  >
    Label
  </label>
</div>
`;
